components/num-util/src/fixed_decimal.rs

```diff
@@ -373,6 +377,135 @@ impl fmt::Display for FixedDecimal {
+                        }
+                    }
+                    true => {
+                        return Err(Error::Syntax);
+                    }
+                }
+            } else if !c.is_digit(10) {
+                return Err(Error::Syntax);
+            }
+        }
+        // no_dot_str: shows the string when the dot is removed from it
+        let no_dot_str = no_sign_str.replace(".", "");
+        let no_dot_str_len = no_dot_str.chars().count();
+
+        // left_zeros: number of zeros at the left of no_dot_str
+        // right_zeros: number of zeros at the right of no_dot_str
+        // These are not necessarily leading or trailing zeros
+        // Example:
+        //     input string     left_zeros      right_zeros
+        //     00123000             2                3
+        //     0.0123000            2               3
+        //     001.23000            2               3
+        //     001230.00            2               3
+        let mut left_zeros = 0;
+        for (_i, c) in no_dot_str.chars().enumerate() {
+            if c == '0' {
+                left_zeros += 1;
+            } else {
+                break;
+            }
+        }
+        // If the input only has 0 (like 000, 00.0, -00.000) we handle the situation here
+        // and return the correct FixedDecimal now and don't run the rest of the code
+        if left_zeros == no_dot_str_len {
+            let mut dec: FixedDecimal = 0.into();
+            dec.is_negative = is_negative;
+            dec.upper_magnitude = left_zeros as i16 - 1i16;
+            if has_dot {
+                dec.upper_magnitude -= no_dot_str_len as i16 - dot_index as i16;
+                dec.lower_magnitude -= no_dot_str_len as i16 - dot_index as i16;
+            }
+            return Ok(dec);
+        }
+        let mut right_zeros = 0;
+        for (_i, c) in no_dot_str.chars().rev().enumerate() {
+            if c == '0' {
+                right_zeros += 1;
+            } else {
+                break;
+            }
+        }
+        // digits_str: the string after removing sign, dot, left zeros, and right zeros
+        let digits_str = &no_dot_str[left_zeros..no_dot_str_len - right_zeros];
+
+        // shift: number of digits after dot in the input_str, 0 if there is no dot
+        let mut shift: i16 = 0;
+        if has_dot {
+            shift = (no_sign_str_len - dot_index - 1) as i16;
+        }
+
+        // Constructing DecimalFixed only using digits_str
+        let mut dec: FixedDecimal = 0.into();
+        let mut v = SmallVec::<[u8; 8]>::new();
+        for (i, c) in digits_str.chars().enumerate() {
+            if i > std::i16::MAX as usize {
+                return Err(Error::Limit);
+            }
+            v.push(c as u8 - '0' as u8);
+        }
+        let v_len = v.len();
+        dec.digits = v;
+        dec.magnitude = v_len as i16 - 1;
+        dec.upper_magnitude = v_len as i16 - 1;
+        dec.lower_magnitude = 0;
```

Comment:
You can set `dec.upper_magnitude` and `dec.lower_magnitude` using the same formulas as in my comment above.  In fact, consider making your output `dec` as soon as you've found the dot position, and then if the string is all zeros, just return it.

For `dec.magnitude`, try this:

```rust
// We can cast with "as" here because lower and upper magnitude are in range
dec.magnitude = ((dot_index as i32) - (left_zeros as i32) - 1i32) as i16;
```

Maybe you can clean that up a bit.  It should work including in cases like "000.001" where the magnitude ends up being negative.

If you do this, you won't need "shift" any more.